### PR TITLE
feat: Multiple OAuth providers

### PR DESCRIPTION
**Pull request by jachym-tousek-keboola**
Jira: https://keboola.atlassian.net/browse/PSGO-409

**Changes:**
- Added support for multiple OAuth providers

-----------

### CODE REVIEW

File: internal/pkg/service/appproxy/http/router.go [hunk 1]
```diff
@@ -57,11 +91,251 @@ func (r *Router) CreateHandler() http.Handler {
...
+		provider.proxy.ServeHTTP(writer, newRequest)
+	})
+
+	// Request to the data app itself
+	handler.HandleFunc("/", func(writer http.ResponseWriter, request *http.Request) {
+		var provider oauthProvider
+		ok := false
+
+		// Identify the provider chosen by the user using a cookie
+		cookie, err := request.Cookie(ProviderCookie)
```
[the remark this answers is not in the document]
Comment (hosekpeter) [reply internal/pkg/service/appproxy/http/router.go:233]: but I think we can log that, then we wont know what happened.

File: internal/pkg/service/appproxy/http/router.go [hunk 2]
```diff
@@ -2,48 +2,81 @@ package http
...
+	config            config.Config
+	clock             clock.Clock
+	handlers          map[AppID]http.Handler
+	selectionTemplate *template.Template
 }
 
-func NewRouter(ctx context.Context, d dependencies.ServiceScope, apps []DataApp) *Router {
+const ProviderCookie = "_oauth2_provider"
+
+const selectionPagePath = "/proxy/selection"
```
Comment (michaljurecko) [inline internal/pkg/service/appproxy/http/router.go:40]: `/proxy/selection` -> `/_proxy/selection`

File: internal/pkg/service/appproxy/http/router.go [hunk 3]
```diff
@@ -57,11 +90,248 @@ func (r *Router) CreateHandler() http.Handler {
...
+
+// OAuth2 Proxy doesn't support multiple providers despite the possibility of setting them up in configuration.
+// So instead we're using separate proxy instance for each provider with a cookie to remember the selection.
+// See https://github.com/oauth2-proxy/oauth2-proxy/issues/926
+func (r *Router) createMultiProviderHandler(oauthProviders map[string]oauthProvider) http.Handler {
+	handler := http.NewServeMux()
+
+	// Request to provider selection page
+	handler.HandleFunc(selectionPagePath, func(writer http.ResponseWriter, request *http.Request) {
+		selection := request.URL.Query().Get("select")
```
Comment (michaljurecko) [inline internal/pkg/service/appproxy/http/router.go:176]: `/selection?select=1` :thinking: maybe `/selection?provider=1` ?

File: internal/pkg/service/appproxy/http/writer.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	"bufio"
+	"net"
+	"net/http"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+type StatusCodeCallback func(writer http.ResponseWriter, statusCode int)
+
+// callbackResponseWriter implements http.ResponseWriter and allows adjusting the response right before the status code is written.
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/appproxy/http/writer.go:13]: Ok.

[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/appproxy/http/writer.go:13]: I would like it if there was an easier way to do this. But this is ok for me.

**Review by michaljurecko — APPROVED** [review]
LGTM

**Pull request merged**